src/connection/transport/sasl.rs

```diff
@@ -30,19 +46,100 @@ impl Credentials {
 }
 
 impl SaslConfig {
-    pub(crate) fn credentials(&self) -> Credentials {
+    pub(crate) async fn get_sasl_config(&self) -> Result<Arc<SASLConfig>, SaslError> {
         match self {
-            Self::Plain(credentials) => credentials.clone(),
-            Self::ScramSha256(credentials) => credentials.clone(),
-            Self::ScramSha512(credentials) => credentials.clone(),
+            Self::Plain(credentials)
+            | Self::ScramSha256(credentials)
+            | Self::ScramSha512(credentials) => Ok(SASLConfig::with_credentials(
+                None,
+                credentials.username.clone(),
+                credentials.password.clone(),
+            )?),
+            Self::Oauthbearer(credentials) => {
+                // Fetch the token first, since that's an async call.
+                // The user can and should cache the token as appropriate to their OAUTH provider.
```

Comment:
```rust
// The user can and should cache the token as appropriate to their OAUTH provider.
```

Could you move this particular note into the docstring for `OauthCallback`? I think this is important for the user.